### ISSUE TITLE
Rule Change: Support simple regex in no-restricted-imports group

### ISSUE THREAD

**Issue by ArielGueta** [eslint/eslint#18536]
### What rule do you want to change?

no-restricted-imports

### What change do you want to make?

Support regex in group

### How do you think the change should be implemented?

```
 "group": "myRegex"
```

### Example code
Currently, the use of the group property can be cumbersome and challenging. For instance:

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    "importNamePattern": "_Enum$",
    "group": ["@app/api", "!@app/api/", "@app/api/**", "!@app/api/enums"]
}]}]*/

// error
import { Foo_Enum } from '@app/api';
import { Bar_Enum } from '@app/api/bar';
import { Baz_Enum } from '@app/api/baz';

// no error
import { B_Enum } from '@app/api/enums';
```

### What does the rule currently do for this code?

Only allows to import the pattern from `@app/api/enums`

### What will the rule do after it's changed?

```
 "group": "myRegex"
```

Allow passing a simple regex.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by Tanujkanti4441** [comment]
Hi @ArielGuetam, thanks for the issue, seems like a good idea to me to expand the `group` option with regex string for reporting imports.

@eslint/eslint-team, should we consider this?

**Comment by zernie** [comment]
Agreed, globs are too limiting

**Comment by nzakas** [comment]
I'm not sure I understand why this is needed. The example given in the description doesn't really explain what the limitations are or how allowing arbitrary regular expressions would solve it. A clearer example along with an actual description of the problem would help make this discussion go smoother.

**Comment by Tanujkanti4441** [comment]
based on issue [#18454](https://github.com/eslint/eslint/issues/18454), i think the problem isn't the limitation of `gitignore-style` but regular expression would be easier to specify a particular module to restrict without specify it's subdirectories or parent directories.

**Comment by nzakas** [comment]
@Tanujkanti4441 easier in what way? As I said in my previous comment, we need to see some actual examples here.

**Comment by Tanujkanti4441** [comment]
```js
eslint no-restricted-imports: ["error", { patterns: [{
    "importNamePattern": "_Enum$",
    "group": ["@app/api", "!@app/api/"]
}]}]

// error
import { Foo_Enum } from '@app/api';

// no error
import { Foo_Enum } from '@app/api/bar';
import { Foo_Enum } from '@app/api/baz';
```
In above example we have to use two patterns to restrict a single `@app/api` but with regex `^@app/api$` pattern can do the same with a single pattern.

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    "importNamePattern": "_Enum$",
    "group": ["@app/api", "!@app/api/", "@app/api/**", "!@app/api/enums"]
}]}]*/

// error
import { Foo_Enum } from '@app/api';
import { Bar_Enum } from '@app/api/bar';
import { Baz_Enum } from '@app/api/baz';

// no error
import { B_Enum } from '@app/api/enums';
```
In this example the pattern inside `group` does not report `@app/api/enums` but reports `@app/api/enums/foo`, to prevent it from being reported also we have to add another pattern after `!@app/api/enums` 
[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1pbXBvcnRzOiBbXCJlcnJvclwiLCB7IHBhdHRlcm5zOiBbe1xuICAgIFwiaW1wb3J0TmFtZVBhdHRlcm5cIjogXCJfRW51bSRcIixcbiAgICBcImdyb3VwXCI6IFtcIkBhcHAvYXBpXCIsIFwiIUBhcHAvYXBpL1wiLCBcIkBhcHAvYXBpLyoqXCIsIFwiIUBhcHAvYXBpL2VudW1zXCIsICFAYXBwL2FwaS9lbnVtcy8qKiwgXVxufV19XSovXG5cbi8vIGVycm9yXG5pbXBvcnQgeyBGb29fRW51bSB9IGZyb20gJ0BhcHAvYXBpJztcbmltcG9ydCB7IEJhcl9FbnVtIH0gZnJvbSAnQGFwcC9hcGkvYmFyJztcbmltcG9ydCB7IEJhel9FbnVtIH0gZnJvbSAnQGFwcC9hcGkvYmF6JztcblxuLy8gbm8gZXJyb3JcbmltcG9ydCB7IEJfRW51bSB9IGZyb20gJ0BhcHAvYXBpL2VudW1zL2Zvbyc7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==)

but same action can be done with a single pattern regex pattern `^@app\/api(\/(?!enums\b)[^\/]+)*$`.

**Comment by nzakas** [comment]
Thanks, that's helpful. I can see how this would be useful.

@eslint/eslint-team other opinions on this?

**Comment by mdjermanovic** [comment]
I'm fine with adding regular expressions. 

gitignore patterns are specifically designed for matching relative file paths, while module specifiers can generally be anything (e.g., URL, host-specific protocol...) so for a runtime-agnostic core rule it makes sense to provide a more generic matching utility like regex patterns.

**Comment by fasttime** [comment]
I'm also in favor of adding an option to match import module specifiers against a regular expression. Just not sure if it makes sense to make `patterns[index].group` accept a regex pattern as suggested by the OP, or if it's better to provide a new option.

**Comment by nzakas** [comment]
Marking as accepted.

> Just not sure if it makes sense to make patterns[index].group accept a regex pattern as suggested by the OP, or if it's better to provide a new option.

I think it needs to be a separate option that accepts regular expression strings (not regular expression objects). We are still trying to keep core rule options as serializable. Maybe `regexGroup`?

**Comment by snitin315** [comment]
I'll work on it.
